File: Python/textbook_track/chapter09/ba9q.py
# Construct the Partial Suffix Array of a String
import itertools as it
import sys

from collections import Counter, defaultdict


def sort_single_characters(text):
    indices = defaultdict(list)
    for ix, char in enumerate(text):
        indices[char].append(ix)

    unique_chars = sorted(indices.keys())
    char_indices = (indices[char] for char in unique_chars)
    return tuple(it.chain.from_iterable(char_indices))


def compute_character_classes(text, order, n):
    classes = [0] * n
    label = 0
    classes[order[0]] = label
    for o1, o2 in zip(order, order[1:]):
        if text[o1] != text[o2]:
            label += 1
        classes[o2] = label
    return tuple(classes)


def count_cumulative_class_sizes(classes, n):
    counts = [0] * n
    class_sizes = Counter(classes)
    for class_id, class_size in class_sizes.items():
        counts[class_id] = class_size

    for ix in range(1, n):
        counts[ix] += counts[ix-1]

    return counts


def sort_doubled_shifts(order, classes, n, cyclic_shift_size):
    counts = count_cumulative_class_sizes(classes, n)
    new_order = [0] * n
    for ix in range(n-1, -1, -1):
        start = (order[ix] - cyclic_shift_size) % n
        class_id = classes[start]
        counts[class_id] -= 1
        new_order[counts[class_id]] = start
    return tuple(new_order)


def update_classes(order, classes, cyclic_shift_size, n):
    new_classes = [0] * n
    label = 0
    new_classes[order[0]] = label
    for current, previous in zip(order[1:], order):
        mid = (current + cyclic_shift_size) % n
        mid_previous = (previous + cyclic_shift_size) % n
        if classes[current] != classes[previous] or classes[mid] != classes[mid_previous]:
            label += 1
        new_classes[current] = label
    return tuple(new_classes)


def compute_full_suffix_array(text):
    n = len(text)
    order = sort_single_characters(text)
    classes = compute_character_classes(text, order, n)
    cyclic_shift_size = 1
    while cyclic_shift_size < n:
        order = sort_doubled_shifts(order, classes, n, cyclic_shift_size)
        classes = update_classes(order, classes, cyclic_shift_size, n)
        cyclic_shift_size *= 2
    return order


def compute_partial_suffix_array(text, k):
    suffix_array = compute_full_suffix_array(text)
    return list(filter(lambda x: x[1] % k == 0, enumerate(suffix_array)))


def main():
    reader = sys.stdin
    text = next(reader).rstrip()
    k = int(next(reader))
    result = compute_partial_suffix_array(text, k)
    for ix, suffix_ix in result:
        print(f'{ix},{suffix_ix}')


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9Q.scala
package TextbookTrack.Chapter09

object BA9Q {
  import scala.annotation.tailrec
  import TextbookTrack.RichNumber._

  private def sortSingleCharacters(text: String): (Vector[Int], Vector[Int]) = {
    val indices: Map[Char, IndexedSeq[Int]] = text.zipWithIndex.groupMap{ case (char, _) => char }{ case (_, ix) => ix }
    val uniqueChars: Vector[Char] = indices.keys.toVector.sorted
    val charIds: Map[Char, Int] = uniqueChars.zipWithIndex.toMap
    val order: Vector[Int] = uniqueChars.flatMap(indices(_))
    val classes: Vector[Int] = text.map(charIds(_)).toVector
    (order, classes)
  }

  private def countCumulativeClassSizes(classes: Vector[Int], n: Int): Array[Int] = {
    val cumulativeCounts: Array[Int] = Array.fill(n)(0)
    val classSizes: Map[Int, Int] = classes.groupMapReduce(identity)(_ => 1)(_ + _)
    classSizes.foreach{ case (classId, classSize) => cumulativeCounts(classId) = classSize }
    (1 until n).foreach(ix => cumulativeCounts(ix) += cumulativeCounts(ix - 1))
    cumulativeCounts
  }

  private def sortDoubledShifts(order: Vector[Int], classes: Vector[Int], n: Int, cyclicShiftSize: Int): Vector[Int] = {
    val counts: Array[Int] = countCumulativeClassSizes(classes, n)
    val newOrder: Array[Int] = Array.fill(n)(0)
    for { ix <- (n - 1) to 0 by -1 } {
      val start: Int = (order(ix) - cyclicShiftSize) mod n
      val classId: Int = classes(start)
      counts(classId) -= 1
      newOrder(counts(classId)) = start
    }
    newOrder.toVector
  }

  private def updateClasses(order: Vector[Int], classes: Vector[Int], cyclicShiftSize: Int, n: Int): Vector[Int] = {
    val newClasses: Array[Int] = Array.fill(n)(0)
    newClasses(order(0)) = 0
    val orders: Iterator[(Int, Int)] = order.sliding(2).collect{ case Vector(a, b) => (a, b) }

    @tailrec
    def loop(label: Int): Unit = {
      if (orders.isEmpty) ()
      else {
        val (previous, current): (Int, Int) = orders.next()
        val mid: Int = (current + cyclicShiftSize) % n
        val midPrevious: Int = (previous + cyclicShiftSize) % n
        val nextLabel: Int =
          if (classes(current) != classes(previous) || classes(mid) != classes(midPrevious)) label + 1 else label
        newClasses(current) = nextLabel
        loop(nextLabel)
      }
    }

    loop(0)
    newClasses.toVector
  }

  private def computeFullSuffixArray(text: String): Vector[Int] = {
    val n: Int = text.length
    val (order, classes): (Vector[Int], Vector[Int]) = sortSingleCharacters(text)

    @tailrec
    def loop(cyclicShiftSize: Int, cyclicShiftOrders: Vector[Int], equivalenceClasses: Vector[Int]): Vector[Int] = {
      if (cyclicShiftSize >= n) cyclicShiftOrders
      else {
        val doubleShiftOrders: Vector[Int] =
          sortDoubledShifts(cyclicShiftOrders, equivalenceClasses, n, cyclicShiftSize)
        val updatedClasses: Vector[Int] = updateClasses(doubleShiftOrders, equivalenceClasses, cyclicShiftSize, n)
        loop(2 * cyclicShiftSize, doubleShiftOrders, updatedClasses)
      }
    }

    loop(1, order, classes)
  }

  def computePartialSuffixArray(text: String, k: Int): List[(Int, Int)] = {
    val suffixArray: Vector[Int] = computeFullSuffixArray(text)
    suffixArray
      .iterator
      .zipWithIndex
      .withFilter{ case (suffixIx, _) => suffixIx % k == 0 }
      .map(_.swap)
      .toList
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val k: Int = reader.next().toInt
    val result: List[(Int, Int)] = computePartialSuffixArray(text, k)
    result.foreach{ case (ix, suffixIx) => println(s"$ix,$suffixIx") }
  }
}
